### PR TITLE
Add an example showing how the parser can control the lexer's mode.

### PR DESCRIPTION
**Pull request by nwalfield**
In some languages, it is necessary for the parser to control the lexer's mode.  This is possible, but non-obvious.

This commit adds a simple example that shows how to parse the "list of length-value" language in which values are encoded as a length, followed by a colon, and then the literal number of bytes, e.g., `2:hi`.  To parse this, the parser need to tell the lexer to return the next `n` literal bytes when it sees the length prefix.

See #802.

### CODE REVIEW

**Comment by nwalfield** [comment]
quick format done.  I also updated the commit message to reference #706.

File: doc/lexer-modes/src/grammar.lalrpop
Comment (dburgener) [inline doc/lexer-modes/src/grammar.lalrpop:62]: I don't know that it really matters too much for the purpose of documentation, but is there a reason to lex the counts by digit and then reconstruct numbers at the parser level rather than when the lexer isn't in literal mode just taking all the digits up to the colon as a single token?

Comment (nwalfield) [reply doc/lexer-modes/src/grammar.lalrpop:62]: This example is a distilled version of a [sexp parser](https://datatracker.ietf.org/doc/html/draft-rivest-sexp-09), [which I wrote](https://gitlab.com/sequoia-pgp/sequoia/-/commit/464d289daf0888cd4eeaa02a23ec1cbbecbd5686).  As far as I can tell, when parsing sexp expressions, the lexer doesn't have enough information to decide how to parse the digits.  In this example, it does, and if you want to change it, that's fine by me.

File: doc/lexer-modes/Cargo.toml
Comment (nwalfield) [reply doc/lexer-modes/Cargo.toml:8]: The other examples have the same issue:

```
$ grep 'lalrpop =' $(find -name Cargo.toml)
./calculator/Cargo.toml:lalrpop = { version = "0.20.2", path = "../../lalrpop" }
./pascal/lalrpop/Cargo.toml:lalrpop = { version = "0.20.2", path = "../../../lalrpop" }
./nobol/Cargo.toml:lalrpop = { version = "0.20.2", path = "../../lalrpop" }
./lexer-modes/Cargo.toml:lalrpop = { version = "0.20.2", path = "../../lalrpop" }
./lexer/Cargo.toml:lalrpop = { version = "0.20.2", path = "../../lalrpop" }
```

Comment (dburgener) [reply doc/lexer-modes/Cargo.toml:8]: My point is that [an open PR](https://github.com/lalrpop/lalrpop/pull/903/files) bumps these numbers for all the other examples.  Whichever of this PR or #903 merges second will need to update to sync up, since #903 obviously doesn't currently bump them for this.

File: doc/lexer-modes/src/grammar.lalrpop
Comment (dburgener) [reply doc/lexer-modes/src/grammar.lalrpop:62]: Okay, I'm fine with leaving it as is.  We can always update it post-merge if desired.

**Comment by dburgener** [comment]
The rust code is hitting some clippy warnings, so you'll need to clean those up before merge.

File: doc/lexer-modes/build.rs
Comment (Pat-Lafon) [inline doc/lexer-modes/build.rs:2]: Pinging #911

**Review by Pat-Lafon — APPROVED** [review]
This looks good to me. Since it's a documentation change, I think the benefit would be to merge this.

File: doc/lexer-modes/src/lexer.rs
Comment (yannham) [inline doc/lexer-modes/src/lexer.rs:48]: Nitpick: beside history, I don't see a good reason why token should be spelled in upper case.

**Review by yannham — APPROVED** [review]

File: doc/lexer-modes/src/lexer.rs
Comment (nwalfield) [reply doc/lexer-modes/src/lexer.rs:48]: In Sequoia, we choose this convention to make the distinction between tokens and productions in the grammar clearer.

Comment (nwalfield) [reply doc/lexer-modes/src/lexer.rs:48]: To be clear: I'm not saying that you or anyone else should adopt this convention. I'm explaining that I have a good reason to use this convention.

**Review by dburgener — APPROVED** [review]

File: doc/lexer-modes/src/lexer.rs
Comment (yannham) [reply doc/lexer-modes/src/lexer.rs:48]: But you need to write a mapping in the `grammar.lalrpop`, don't you? So you could do:

```
        LPAREN => lexer::Token::LParen,
```

And keep the Rust-side camel case. Anyway, as I said, it's a nitpick - I'm fine keeping it like that, and I surely don't want to block this PR on that :slightly_smiling_face:

Comment (nwalfield) [reply doc/lexer-modes/src/lexer.rs:48]: I could, but I don't 🙂.  This isn't my project. I'm happy for you to modify the code to suit your tastes.

Comment (yannham) [reply doc/lexer-modes/src/lexer.rs:48]: Sure, it's just that

> In Sequoia, we choose this convention to make the distinction between tokens and productions in the grammar clearer.

My point was that you can still keep the distinction between tokens and productions in the grammar clear (as in the example above), while using camel case for the Rust enum. As often, I suspect the actual answer is mostly that it's a matter of taste, and that's ok :slightly_smiling_face: 

Honestly this is quite unimportant, so let's merge like this. Thanks again for the contribution!

**Comment by Pat-Lafon** [comment]
Before I go wrongly closing a bunch of issues, I'll just list them here in case I'm mistaken(or if anyone sees those issues since this pr might help them solve their issue) : #706, #195, #673, #802

**Comment by dburgener** [comment]
My reading of #706, #195, and #673 is that they are more about implementing this as a direct feature in lalrpop's lexer.  The approach here requires a custom lexer, and then the parser can exchange state between itself and that custom lexer to control behavior.  I haven't understood #673 at any sort of real level, but my impression is that the high level approach there is to make the built-in lexer aware of what tokens would be permitted at a given parse state, and then disambiguate lexing using that information.

So I would say that those three seem to be fundamentally about achieving this *without* a custom lexer, which is still a feature that lalrpop does not have today.

#802 does seem to me to just be a question of "is this achievable at all?" and the answer provided by this PR is "yes, and here's an example of how to do it using a custom lexer".  So I'd think that #802 can be closed with a pointer to this example.

**Pull request merged**
